### PR TITLE
KAFKA-15996: Improve JsonConverter performance

### PR DESCRIPTION
**Pull request by mfvitale**
https://issues.apache.org/jira/browse/KAFKA-15996

After some performance tests in our Debezium connectors we have found some bottleneck in serdes using JsonConverter. 

JFR without Afterburner module
image

JFR Afterburner module
image

Seems that the speed up is more then 20% that is not so bad.

I have followed [Jackson Performance](https://github.com/FasterXML/jackson-docs/wiki/Presentation:-Jackson-Performance).

I saw also that there is the [Blackbird](https://github.com/FasterXML/jackson-modules-base/blob/master/blackbird/README.md) that is for Java 11+ Environment. Should we use it?

### CODE REVIEW

**Comment by mfvitale** [comment]
@C0urante @mimaison can you please gently have a look?

**Comment by mimaison** [comment]
Thanks @mfvitale for the PR. This looks like a significant performance improvement. We can't upgrade to Blackbird just yet as Kafka 3.X still supports Java 8.

Could you write a small jmh benchmark to demonstrate the performance difference? For example, see https://github.com/apache/kafka/pull/13776

**Comment by mfvitale** [comment]
> Could you write a small jmh benchmark to demonstrate the performance difference? For example, see #13776

@mimaison , Done! These are some results

```text
Benchmark                                   (afterBurnModule)    Mode      Cnt        Score    Error  Units
JsonConverterBenchmark.deserialize                       true  sample  1387169    17997.199 ± 17.127  ns/op
JsonConverterBenchmark.deserialize:p0.00                 true  sample             16384.000           ns/op
JsonConverterBenchmark.deserialize:p0.50                 true  sample             17408.000           ns/op
JsonConverterBenchmark.deserialize:p0.90                 true  sample             18304.000           ns/op
JsonConverterBenchmark.deserialize:p0.95                 true  sample             20672.000           ns/op
JsonConverterBenchmark.deserialize:p0.99                 true  sample             30720.000           ns/op
JsonConverterBenchmark.deserialize:p0.999                true  sample             40757.120           ns/op
JsonConverterBenchmark.deserialize:p0.9999               true  sample            423056.896           ns/op
JsonConverterBenchmark.deserialize:p1.00                 true  sample            924672.000           ns/op
JsonConverterBenchmark.deserialize                      false  sample  1303600    19164.929 ± 21.371  ns/op
JsonConverterBenchmark.deserialize:p0.00                false  sample             16576.000           ns/op
JsonConverterBenchmark.deserialize:p0.50                false  sample             18048.000           ns/op
JsonConverterBenchmark.deserialize:p0.90                false  sample             21152.000           ns/op
JsonConverterBenchmark.deserialize:p0.95                false  sample             24736.000           ns/op
JsonConverterBenchmark.deserialize:p0.99                false  sample             37376.000           ns/op
JsonConverterBenchmark.deserialize:p0.999               false  sample             49472.000           ns/op
JsonConverterBenchmark.deserialize:p0.9999              false  sample            450703.258           ns/op
JsonConverterBenchmark.deserialize:p1.00                false  sample           1052672.000           ns/op
JsonConverterBenchmark.serialize                         true  sample  1489235     8396.143 ± 12.427  ns/op
JsonConverterBenchmark.serialize:p0.00                   true  sample              7296.000           ns/op
JsonConverterBenchmark.serialize:p0.50                   true  sample              8072.000           ns/op
JsonConverterBenchmark.serialize:p0.90                   true  sample              8640.000           ns/op
JsonConverterBenchmark.serialize:p0.95                   true  sample              9552.000           ns/op
JsonConverterBenchmark.serialize:p0.99                   true  sample             15456.000           ns/op
JsonConverterBenchmark.serialize:p0.999                  true  sample             23520.000           ns/op
JsonConverterBenchmark.serialize:p0.9999                 true  sample             38724.890           ns/op
JsonConverterBenchmark.serialize:p1.00                   true  sample           1159168.000           ns/op
JsonConverterBenchmark.serialize                        false  sample  1411941     8850.354 ± 12.808  ns/op
JsonConverterBenchmark.serialize:p0.00                  false  sample              7552.000           ns/op
JsonConverterBenchmark.serialize:p0.50                  false  sample              8336.000           ns/op
JsonConverterBenchmark.serialize:p0.90                  false  sample              9456.000           ns/op
JsonConverterBenchmark.serialize:p0.95                  false  sample             11744.000           ns/op
JsonConverterBenchmark.serialize:p0.99                  false  sample             18464.000           ns/op
JsonConverterBenchmark.serialize:p0.999                 false  sample             25313.856           ns/op
JsonConverterBenchmark.serialize:p0.9999                false  sample             56192.000           ns/op
JsonConverterBenchmark.serialize:p1.00                  false  sample            952320.000           ns/op

```

Others interesting results from [jmv serializers benchmark](https://github.com/eishay/jvm-serializers/wiki)

File: build.gradle
Comment (divijvaidya) [inline build.gradle]: Please help me understand why did we add this dependency in clients?

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java
Comment (divijvaidya) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java]: nit

please maintain original indentation

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java
Comment (divijvaidya) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:248]: please add a javadoc comment specifying that this ctor is visible only for benchmarking. we add similar comments when we expose a new ctor/method for testing.

**Review by divijvaidya — COMMENTED** [review]
Thank you for this PR. I have left some minor comments.

Also, please add a JIRA targeting 4.0 (where we will deprecate JDK 8) to replace afterburn with blackbird.

File: build.gradle
Comment (mfvitale) [reply build.gradle]: I have been misled by other Jackson dependencies. Removed.

**Comment by mfvitale** [comment]
> Thank you for this PR. I have left some minor comments.
> 
> Also, please add a JIRA targeting 4.0 (where we will deprecate JDK 8) to replace afterburn with blackbird.

@divijvaidya https://issues.apache.org/jira/browse/KAFKA-16041

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java
Comment (divijvaidya) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java]: please follow original indentation.

**Comment by divijvaidya** [comment]
@mfvitale one last comment about fixing indentation, otherwise we should be good to ship!

**Review by mimaison — APPROVED** [review]
LGTM, but yes please undo the indentation changes in JsonConverter as suggested by Divij

**Comment by mfvitale** [comment]
@divijvaidya Fixed!

**Review by divijvaidya — APPROVED** [review]
haha! I like your commit message "code beauty" :) 

I see that this is your first change to Apache Kafka. Congratulations and welcome! As a next step, I will wait for CI to be stable before I merge this change in.

**Comment by mfvitale** [comment]
> haha! I like your commit message "code beauty" :) 

😉

> I see that this is your first change to Apache Kafka. Congratulations and welcome! 

Yes! Thanks!

> As a next step, I will wait for CI to be stable before I merge this change in.

Perfect!

**Comment by divijvaidya** [comment]
A fix to make CI stable was pushed today. @mfvitale  Can you please rebase from trunk one time? It will probably help us get a better picture of CI failures/success.

**Comment by mfvitale** [comment]
> A fix to make CI stable was pushed today. @mfvitale Can you please rebase from trunk one time? It will probably help us get a better picture of CI failures/success.

@divijvaidya done.

**Comment by divijvaidya** [comment]
I have verified that failing tests in the CI are known to be flaky or/and are failing for other PRs as well.

- org.apache.kafka.streams.integration.StandbyTaskEOSMultiRebalanceIntegrationTest.shouldHonorEOSWhenUsingCachingAndStandbyReplicas has been failing with similar error message in rest of the PRs as well.

- kafka.server.BrokerRegistrationRequestTest.testRegisterZkWith33Controller failing for other PRs as well

**Comment by mfvitale** [comment]
@divijvaidya do you know if it will be in 3.7?

**Comment by divijvaidya** [comment]
@mfvitale No, we missed that train. We are currently past the code freeze for 3.7 (see: https://cwiki.apache.org/confluence/display/KAFKA/Release+Plan+3.7.0). Given this is a performance improvement feature and not a major bug fix, I am currently not in favour of pushing this to be included as an exception to 3.7. However, if you think that this should go in 3.7, we can discuss about that.

What do you think?

**Comment by mfvitale** [comment]
@divijvaidya I agree this is not a critical issue. Do you think there can be a chance for 3.7.1? Is there any schedule?

**Comment by divijvaidya** [comment]
> Do you think there can be a chance for 3.7.1? Is there any schedule?

We don't usually backport performance fixes to patch versions. (With the usual caveat that this is not set in stone and is open to discussion). Our motivation is to backport only critical changes in the patch version so that we don't accidentally introduce a regression.

We don't current have a date for 3.7.1 but going by last few releases, it is highly likely that we will have a patch 3.7.1 release.

**Comment by mfvitale** [comment]
> We don't usually backport performance fixes to patch versions. (With the usual caveat that this is not set in stone and is open to discussion). Our motivation is to backport only critical changes in the patch version so that we don't accidentally introduce a regression.

It's fair. 

So, normally it's expected to be in 3.8 release. Do you have any estimations?

**Comment by divijvaidya** [comment]
> Do you have any estimations?

3.8 is usually expected to release around ~Apr-May timeframe. You can find more details about timelines of our releases here: https://cwiki.apache.org/confluence/display/KAFKA/Time+Based+Release+Plan

**Pull request merged**
